### PR TITLE
Rework calendar screen

### LINKED ISSUE #175: Event search view shows old events first
### Describe the bug
The search view for events shows old events first

image

### How to reproduce
1. Go to the Calendar
2. Search for an event (e.g. Weekend)

### Expected behaviour
Show new events first.

### PR DESCRIPTION
**Pull request by JAicewizard**
Closes #175 .

### Summary
This PR does a couple things:
 - make the search and normal view consistent
 - add option to go up

This also means search no longer starts at the beginning

### How to test
Steps to test the changes you made:
1. Go to calendar
2. Click on search
3. Click on load more (probably pull in the future)

Still to do 
- [x] pull to load more
- [x] auto refresh after a while (only on load)
- [x] "today" button
- [x] add way to load more in the past
- [x] add dummy day if it doesnt exist for today, and make the date magenta

### CODE REVIEW

File: lib/blocs/calendar_cubit.dart [hunk 1]
```diff
@@ -271,11 +280,77 @@ class CalendarCubit extends Cubit<CalendarState> {
...
+      }
+
+      emit(CalendarState.success(
+          results: events, isDone: isDone, isDoneUp: oldState.isDoneUp));
+    } on ApiException catch (exception) {
+      emit(CalendarState.failure(message: exception.message));
+    }
+  }
+
+  Future<void> moreup() async {
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:294]: Could you rename to `moreUp`?

File: lib/blocs/list_state.dart [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import 'package:equatable/equatable.dart';
 
+//TODO should this be split into two, one for up/down and one for just down
```
Comment (DeD1rk) [inline lib/blocs/list_state.dart:3]: Yeah I think so. Would be nice to have e.g. `TwoWayListState extends ListState` although I'm actually not sure whether extending is needed or desirable

File: lib/blocs/list_state.dart [hunk 3]
```diff
@@ -61,22 +72,30 @@ class ListState<T> extends Equatable {
       : message = null,
         isLoading = true,
         isLoadingMore = false,
-        isDone = true;
+        isLoadingMoreUp = false,
+        isDone = true,
+        isDoneUp = true;
 
   const ListState.loadingMore({required this.results})
```
Comment (DeD1rk) [inline lib/blocs/list_state.dart:79]: This constructor would probably cause rendering a loading indicator at both sides, instead of only on the side that is actually loading

Comment (DeD1rk) [reply lib/blocs/list_state.dart:79]: On a side note, any idea if the cubit's `more[Up]()` will still work correctly if both sides are loading? This is a situation that's likely to occur naturally when only 1 page has been loaded so far

File: lib/blocs/calendar_cubit.dart [hunk 4]
```diff
@@ -122,16 +122,22 @@ class CalendarCubit extends Cubit<CalendarState> {
...
   int _nextOffset = 0;
+  int _nextPastOffset = 0;
 
   /// The time used as filter, stored so that later
   /// paginated requests have the correct offset.
   DateTime? _lastLoadTime;
 
+  /// A list of events that have been removed from the previous results
+  /// in order to prevent them filling up the calendar before today.
+  /// These should be added in later calls to [more()].
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:133]: Are these comments swapped?

Comment (DeD1rk) [reply lib/blocs/calendar_cubit.dart:133]: Or actually, `more()` and `moreup()`?

**Review by DeD1rk — COMMENTED** [review]
Sick! There's a few comments, but I think it looks like it should work great. Is there any chance I could convince you to write a unittest for the cubit (using a mocked api)?  ('no' is fine as long as we test some edge cases)
I might take a look at that myself too but I'm not sure when I'll have the time for that so no promises.

File: lib/blocs/list_state.dart (hunk 3 above)
Comment (JAicewizard) [reply lib/blocs/list_state.dart:81]: Uhh, I think thats a datarace, its not really harmful but we should probably prevent loading up if we are loading down

**Comment by JAicewizard** [comment]
I put way to much time into this (but learned a bit as well). It works best in profile mode otherwise the animations are just ~0fps. We might want to put fetching+processing in an isolate if that's possible to avoid jank.

Also long running events like on staging are really annoying to deal with, as we receive them per event but split them up per day. not sure yet why this is a problem when going up not down

**Comment by DeD1rk** [comment]
I suspect the moving up may trigger a rebuild of the items, because the card at each index changes, where loading more beliw only adds new ones without changing the existing items.

**Comment by DeD1rk** [comment]
A solution can be to have separe past and future lists. https://github.com/flutter/flutter/issues/20608 is related

**Comment by JAicewizard** [comment]
I will try testing a bit more, but currently there are not that many events below so I think ill postpone that to later

**Comment by JAicewizard** [comment]
works nice on release bit jank on debug, not sure why

File: lib/blocs/calendar_cubit.dart [hunk 5]
```diff
@@ -122,88 +132,153 @@ class CalendarCubit extends Cubit<CalendarState> {
 
   /// The offset to be used for the next paginated request.
   int _nextOffset = 0;
+  int _nextPastOffset = 0;
+
+  /// The time rounded down to the month used to split the events in "past" and
+  /// "future". Also used in the query. This is a final to avoid race conditions
+  final DateTime _splitTime =
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:139]: Shouldn't this be set in load() (so that force-reloading does use a new split time)?

File: lib/blocs/calendar_cubit.dart [hunk 6]
```diff
@@ -122,88 +132,153 @@ class CalendarCubit extends Cubit<CalendarState> {
...
+
+      // Split multi-day events.
+      final pastEvents = [
+        ...pastPartnerEventsResponse.results
+            .expand(CalendarEvent.splitEventIntoCalendarEvents)
+            .toList(),
+        ...pastEventsResponse.results
+            .expand(CalendarEvent.splitEventIntoCalendarEvents)
+            .toList(),
+      ].toList();
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:265]: This needs to be sorted before filtering out below right?

File: lib/blocs/calendar_cubit.dart [hunk 7]
```diff
@@ -122,88 +132,153 @@ class CalendarCubit extends Cubit<CalendarState> {
...
+      // up the calendar further then where the first not-loaded event will go
+      // later.
+      if (!isDoneUp) {
+        while (pastEvents.isNotEmpty &&
+            (pastEvents.first.parentEvent is PartnerEvent ||
+                pastEvents.first.end != pastEvents.first.parentEvent.end)) {
+          _remainingPastEvents.add(pastEvents.removeAt(0));
         }
       }
+      // Sort only the new events, because the old events in
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:277]: This comment is from a more() thing?

File: lib/blocs/calendar_cubit.dart [hunk 8]
```diff
@@ -122,88 +132,153 @@ class CalendarCubit extends Cubit<CalendarState> {
...
+      final futurePartnerEventsResponseFuture = api.getPartnerEvents(
+        start: _splitTime,
+        search: query,
+        ordering: 'start',
+      );
+      // Get all partner events.
+      final pastPartnerEventsResponseFuture = api.getPartnerEvents(
+        start: _splitTime,
+        search: query,
+        ordering: 'start',
```
Comment (DeD1rk) [inline lib/blocs/calendar_cubit.dart:199]: shouldn't this be on -end? or at least -something, I'm not sure how end/start matters

File: lib/blocs/calendar_cubit.dart (hunk 6 above)
Comment (DeD1rk) [reply lib/blocs/calendar_cubit.dart:265]: This needs to be sorted right?

File: lib/blocs/calendar_cubit.dart (hunk 5 above)
Comment (JAicewizard) [reply lib/blocs/calendar_cubit.dart:139]: I could indeed update it, there, but initializing it here allows us to remove the `?`. Updating it doesnt really matter as it is rounded to the month.

File: lib/api/concrexit_api_repository.dart [hunk 9]
```diff
@@ -86,6 +86,7 @@ class ConcrexitApiRepository implements ApiRepository {
     try {
       final response = await request();
       if (allowedStatusCodes.contains(response.statusCode)) return response;
+      // TODO: add 429 and 500
```
Comment (DeD1rk) [inline lib/api/concrexit_api_repository.dart:89]: Do it? or don't

File: pubspec.yaml [hunk 10]
```diff
@@ -97,7 +97,13 @@ dependencies:
   overlay_support: ^2.1.0
 
   # Sticky headers for calendar.
-  sticky_headers: ^0.3.0+2
+  # wait for https://github.com/letsar/flutter_sticky_header/issues/88 to move to that library
+  # when https://github.com/fluttercommunity/flutter_sticky_headers/pull/72 gets merged we can at least go back to upstream
+  sticky_headers:
+    git:
+      url: https://github.com/JAicewizard/flutter_sticky_headers
+      ref: master # branch name
```
Comment (DeD1rk) [inline pubspec.yaml:105]: Could you fix a sha here?

**Review by DeD1rk — CHANGES_REQUESTED** [review]
Looks about right, very cool. I would love some unit tests to ensure the ordering/filtering logic in the cubit works as it should (in edge cases), because I can't wrap my head around it based on code only.
And I need to test a real build of course

**Comment by JAicewizard** [comment]
@DeD1rk I reworked how the filtering worked as it is indeed a PITA to read, took me a while to know what it did as well. apple upload should work now as well

**Comment by JAicewizard** [comment]
Noticed some issue on release, partner events are not filtered, there are no events in February, and there is a lot of dynamic spacing above februari

**Comment by DeD1rk** [comment]
Stul not working on release :( https://user-images.githubusercontent.com/41264528/220939199-d05f460b-abdf-4ab3-a4db-6103539b4384.MOV

**Comment by DeD1rk** [comment]
Apart from the visual bug, search doesnt filter partner events: image and if i search for an emptystring and pull to refresh, I get this incorrect result: image

**Comment by JAicewizard** [comment]
The visual bug was actually a very critical issue, good catch. I think the partner events is a server bug, check https://thalia.nu/api/v2/partners/events/?search=beest&ordering=start&start=2023-02-01T00%3A00%3A00.000 to see that it actually returns events not matching

**Comment by JAicewizard** [comment]
The partner events are also not filtered in the release of the app, probably not noticed due to them not being very common. I'm gonna say that's not a regression and we can merge with that issue still being present and wait for concrexit to release

**Comment by JeeVee11** [comment]
When searching and then clicking on the cross in the search bar, there is a 'StateError (Bad state: No element)' exeption

**Comment by JAicewizard** [comment]
@JeeVee11 all issues should be fixed now

File: lib/ui/screens/calendar_screen.dart [hunk 11]
```diff
@@ -297,7 +294,9 @@ class CalendarScrollView extends StatelessWidget {
                 SliverToBoxAdapter(
                   child: TextButton(
```
Comment (DeD1rk) [inline lib/ui/screens/calendar_screen.dart:295]: Doesn't need to be a button any more I guess (anyway, as a button its corners when pressed don't align nicely)

**Review by DeD1rk — DISMISSED** [review]
